frappe/core/doctype/doctype/test_doctype.py

```diff
@@ -560,6 +560,48 @@ def test_create_virtual_doctype(self):
 		self.assertEqual(doc.is_virtual, 1)
 		self.assertFalse(frappe.db.table_exists("Test Virtual Doctype"))
 
+	def test_create_virtual_doctype_as_child_table(self):
+		"""Test virtual DocType as Child Table below a normal DocType."""
+		try:
+			frappe.delete_doc("DocType", "Test Parent Virtual DocType", force=1)
+		except Exception:
+			pass
+
+		try:
+			frappe.delete_doc("DocType", "Test Virtual DocType as Child Table", force=1)
+		except Exception:
+			pass
+
+		virtual_doc = new_doctype("Test Virtual DocType as Child Table")
+		virtual_doc.is_virtual = 1
+		virtual_doc.istable = 1
+		virtual_doc.insert(ignore_permissions=True)
+
+		doc = frappe.get_doc("DocType", "Test Virtual DocType as Child Table")
+
+		self.assertEqual(doc.is_virtual, 1)
+		self.assertEqual(doc.istable, 1)
+		self.assertFalse(frappe.db.table_exists("Test Virtual DocType as Child Table"))
+
+		parent_doc = new_doctype("Test Parent Virtual DocType")
+		field_1 = parent_doc.append("fields", {})
+		field_1.fieldname = "virtual_child_table"
+		field_1.fieldtype = "Table"
+		field_1.options = "Test Virtual DocType as Child Table"
+		parent_doc.insert(ignore_permissions=True)
+
+		# create entry for parent doctype
+		parent_doc_entry = frappe.new_doc("Test Parent Virtual DocType")
+		parent_doc_entry.some_fieldname = "Test"
+		parent_doc_entry.insert(ignore_permissions=True)
+
+		# now update the new parent doc
+		parent_doc_entry.some_fieldname = "Test update"
+		parent_doc_entry.save(ignore_permissions=True)
+
+		# now delete the parent doc
+		parent_doc_entry.delete()
```

Comment:
```suggestion
		frappe.delete_doc_if_exists("DocType", "_Test Virtual Parent", force=1)
		frappe.delete_doc_if_exists("DocType", "_Test Virtual Child", force=1)

		virtual_doc = new_doctype("_Test Virtual Child")
		virtual_doc.is_virtual = 1
		virtual_doc.istable = 1
		virtual_doc.append(
			"fields",
			{
				"fieldname": "some_fieldname",
				"fieldtype": "Data",
			},
		)
		virtual_doc.insert(ignore_permissions=True)

		self.assertFalse(frappe.db.table_exists("_Test Virtual Child"))

		parent_doc = new_doctype("_Test Virtual Parent")
		parent_doc.append(
			"fields",
			{
				"fieldname": "virtual_child_table",
				"fieldtype": "Table",
				"options": "_Test Virtual Child",
			},
		)
		parent_doc.insert(ignore_permissions=True)

		# create entry for parent doctype
		parent_doc_entry = frappe.new_doc("_Test Virtual Parent")
		parent_doc_entry.append("virtual_child_table", {"some_fieldname": "Test"})
		parent_doc_entry.insert(ignore_permissions=True)

		# update the new parent doc
		parent_doc_entry.virtual_child_table[0].some_fieldname = "Test update"
		parent_doc_entry.save(ignore_permissions=True)

		# delete the parent doc
		parent_doc_entry.delete()
```

Some ideas how to make the test clearer. The field `some_fieldname` wasn't defined for the parent DocType. Above, I added it as a field to the child DocType. (I didn't check if it actually runs).

By the way, it should be possible to run only the test cases in this file, with the following command:

```
bench run-tests --module frappe.core.doctype.doctype.test_doctype
```

Reply:
Thanks alot for your ideas and especially for `frappe.delete_doc_if_exists` :)

Yes, I can run the module tests, or my single test (and my one is running successful). I was just claiming, that they're not (most of them) are replayable, without resetting the bench's db, as the rollback doesn't work well enough in my setup.

The **some_fieldname** property is created in the method **new_doctype** in the same test file (frappe/core/doctype/doctype/test_doctype.py), so that's why I didn't need to declare it on my own.

I'm just not sure, why you prefer the "_" before the Test DocType names. I can follow your suggestion here for sure; it's just that in the rest of the tests (in this file) none of the testing DocTypes have a leading "_" before it's name. So I just copied most of the instructions from this and other test files, to be as close as possible like the rest.

Reply:
Oh, sorry I didn't check my suggestions properly. Let's just do the changes regarding `delete_doc_if_exists`.

Reply:
I commited a suggestion, with `delete_doc_if_exists` and I also tried to improve the readability based on your other suggestions.

If this is ok for you, it would be great if you could start the tests for a next try again. Otherwise give me a hint of course :)

Thank you so far!